### DISCUSSION TITLE
How to use EMA model to do evaluation or test?

### DISCUSSION THREAD

**Discussion by liming-ai** [open-mmlab/mmdetection#9162] [Q&A] [answered]

**Comment by jbwang1997** [comment]
Do you mean adding EMA to models?  All models trained in mmdetection can do evaluation and testing by tools/test.py

**Comment by liming-ai** [comment]
> Do you mean adding EMA to models? All models trained in mmdetection can do evaluation and testing by tools/test.py

I add custom hook to add EMA during training, but the keys in saved checkpoints is ema_xxx_xxx_xxx, which is hard to transfer to normal key names. I just want to know that is it using EMA model to evaluate results by default if `EMAHook` is in the config during test?

**Comment by jbwang1997** [comment, marked as answer]
In EMA model, we register `ema_xxx_xxx_` in the model as buffer. The normal keys are also in the model state_dict, so you can directly evaluate results by default.

**Comment by liming-ai** [comment]
> In EMA model, we register `ema_xxx_xxx_` in the model as buffer. The normal keys are also in the model state_dict, so you can directly evaluate results by default.

Do you mean directly evaluate results is equal to use the EMA model to evaluate instead of the online model?

**Comment by jbwang1997** [comment]
I mean you can directly load the .pth model in tools/test.py. It will test normally. The parameters starting with `ema` will not influence the work of tools/test.py

**Comment by liming-ai** [comment]
> I mean you can directly load the .pth model in tools/test.py. It will test normally. The parameters starting with `ema` will not influence the work of tools/test.py

Thanks for the answer. My point is, when we using EMA, there are two models, one is online model, and another is the EMA model (the temporal ensemble model). Just like the student model and the teacher model in mean teacher.

I want to know when we evaluate result, which model would be used?

**Comment by jbwang1997** [comment]
You can directly use the saved model. EMAHook will swap the ema model and the normal model before CheckPoint saving.

**Comment by jbwang1997** [comment]
Code can be found at [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/hook/ema.py#L84-L100)

**Comment by liming-ai** [comment]
> Code can be found at [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/hook/ema.py#L84-L100)

Thanks a lot! Could you give me some advice about how to ensemble the results from different detectors?

**Comment by wanghonglie** [comment]
> > Code can be found at [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/hook/ema.py#L84-L100)
> 
> Thanks a lot! Could you give me some advice about how to ensemble the results from different detectors?

Maybe you can refer to more materials related to ensemble learning.
